WIP: Create apiserver cert w/ separate root-ca using cfssl

sttts opened the pull request:
**Based on** https://github.com/kubernetes/kubernetes/pull/37830

- use cfssl to create apiserver self-signed certs with separate root-ca, appended to the actual cert in `apiserver.crt`
- ~~find root-ca for loopback client via `--tls-ca-file`, appended to the server cert or appended to a matching SNI cert~~ also part of https://github.com/kubernetes/kubernetes/pull/39022
- extract root-ca from `apiserver.crt` and write to kubeconfig in `local-cluster-up.sh`

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35143)
<!-- Reviewable:end -->

Commit:
Update cfssl godeps

sttts commented:
@dims rebased. ptal.

dims commented on pkg/util/cert/cert.go:
```diff
@@ -126,56 +134,84 @@ func MakeEllipticPrivateKeyPEM() ([]byte, error) {
+		Profiles: map[string]*config.SigningProfile{},
+		Default:  config.DefaultConfig(),
+	}
+	policy.Default.ExpiryString = fmt.Sprintf("%dh", 24*365*10)
```
Is this 10 years? (i see `24*365*1` in line 172 as a comparison)

sttts replied:
will change it one year as well.

dims commented on pkg/util/cert/cert.go:
```diff
@@ -126,56 +134,84 @@ func MakeEllipticPrivateKeyPEM() ([]byte, error) {
+	caCertPem, _, caKeyPem, err = initca.New(&rootCAReq)
 	if err != nil {
-		return nil, nil, err
+		return nil, nil, nil, nil, fmt.Errorf("error creating root-ca: %v", err)
```
Am guessing end users will see this error, let's spell it out root-ca -> "root certificate authority" if so.

dims commented on pkg/util/cert/cert.go:
```diff
@@ -126,56 +134,84 @@ func MakeEllipticPrivateKeyPEM() ([]byte, error) {
-		template.IPAddresses = append(template.IPAddresses, ip)
-	} else {
-		template.DNSNames = append(template.DNSNames, host)
+	glog.Infof("Creating apiserver key and certificate")
```
"and certificate" to "and certificate signing request". right?

dims commented on pkg/util/cert/cert.go:
```diff
@@ -126,56 +134,84 @@ func MakeEllipticPrivateKeyPEM() ([]byte, error) {
-		return nil, nil, err
+	caCert, err := helpers.ParseCertificatePEM(caCertPem)
+	if err != nil {
+		return nil, nil, nil, nil, fmt.Errorf("failed to parse ca cert: %v", err)
```
"cert" -> "certificate" (if this is a end user visible error)

dims commented on pkg/util/cert/cert.go:
```diff
@@ -126,56 +134,84 @@ func MakeEllipticPrivateKeyPEM() ([]byte, error) {
+	if err != nil {
+		return nil, nil, nil, nil, fmt.Errorf("failed to parse ca cert: %v", err)
+	}
+	glog.Infof("Signing apiserver certificate with root-ca")
```
"root-ca" -> "root certificate authority" (if user visible)

dims commented on pkg/util/cert/cert.go:
```diff
@@ -126,56 +134,84 @@ func MakeEllipticPrivateKeyPEM() ([]byte, error) {
+		Request: string(keyCSR),
+	})
+	if err != nil {
+		return nil, nil, nil, nil, fmt.Errorf("error signing crt: %v", err)
```
crt -> certificate (if user visible)

dims reviewed (commented):
Looks great @sttts Just a couple of questions and a few suggestions

sttts commented on pkg/genericapiserver/config.go:
```diff
@@ -297,6 +300,10 @@ func (c *Config) ApplyOptions(options *options.ServerRunOptions) *Config {
 		}
 
+		if secureServingInfo.ServerCert.Generate && secureServingInfo.ServerCA == "" {
+			secureServingInfo.ServerCA = "ca-" + secureServingInfo.ServerCert.CertFile
```
yes, we have to.

sttts commented:
@dims addressed all comments.

dims commented:
@sttts Looks good! I think we should be all set after fixing the unit test failures.

dims commented:
Oh. looks like https://github.com/kubernetes/kubernetes/pull/36068 has to go in first? (I was wondering where that last commit came from and found that one)

sttts commented:
Fixed the unit test.

dims commented:
@sttts I've looked at the apiserver.crt and ca-apiserver.crt (using `openssl x509 -in apiserver.crt -text -noout`) and they look good. So Looks good to me 👍 

@liggitt Do you want to review this as well? If not i can try "/lgtm" to see if that works for me :)

liggitt commented on pkg/genericapiserver/config.go:
```diff
@@ -282,6 +284,7 @@ func (c *Config) ApplyOptions(options *options.ServerRunOptions) *Config {
 				},
 			},
 			SNICerts: []NamedCertKey{},
+			ServerCA: options.TLSCAFile,
```
doc for `--tls-cert-file` already expects the CA to be concatenated to the cert file:

> --tls-cert-file string
> File containing x509 Certificate for HTTPS. (CA cert, if any, concatenated after server cert)

sttts replied:
Wondering, is that ca concatenated to the cert used anywhere? The https server in the apiserver does not need it.

liggitt commented on pkg/genericapiserver/config.go:
```diff
@@ -190,6 +190,8 @@ type SecureServingInfo struct {
 	// SNICerts are named CertKeys for serving secure traffic with SNI support.
 	SNICerts []NamedCertKey
+	// ServerCA is the certificate bundle needed to verify the ServerCert
+	ServerCA string
```
it's slightly odd for the server to need this... I think the cert file referenced in GeneratableKeyCert or NamedCertKey should already contain the chain

liggitt commented on cmd/kubelet/app/server.go:
```diff
@@ -515,11 +515,16 @@ func InitializeTLS(kc *componentconfig.KubeletConfiguration) (*server.TLSOptions
+				return nil, err
+			}
+
 			if err := certutil.WriteCert(kc.TLSCertFile, cert); err != nil {
```
write CA appended to the cert file instead?

sttts replied:
Yes, I like that. This also solves the logical issue below when to create a new root ca.

sttts replied:
done

liggitt commented on pkg/genericapiserver/config.go:
```diff
@@ -301,6 +304,10 @@ func (c *Config) ApplyOptions(options *options.ServerRunOptions) *Config {
 			})
 		}
 
+		if secureServingInfo.ServerCert.Generate && secureServingInfo.ServerCA == "" {
```
it's also odd for the generate behavior to only trigger when the cert/key args are empty, but to overwrite whatever is specified by a non-empty tls-ca-file arg

dims commented:
@sttts looks like #36068 has merged now. Let's please rebase this and try to get it in. There are some comments too from Jordan

sttts commented:
@dims will work on this later this week, after Kubecon

xilabao commented on hack/local-up-cluster.sh:
```diff
@@ -170,7 +170,7 @@ ENABLE_HOSTPATH_PROVISIONER=${ENABLE_HOSTPATH_PROVISIONER:-"false"}
 ENABLE_CONTROLLER_ATTACH_DETACH=${ENABLE_CONTROLLER_ATTACH_DETACH:-"true"} # current default
 CERT_DIR=${CERT_DIR:-"/var/run/kubernetes"}
-ROOT_CA_FILE=$CERT_DIR/apiserver.crt
+ROOT_CA_FILE=$CERT_DIR/ca-apiserver.crt
```
need to create $CERT_DIR/ca-apiserver.crt

sttts commented:
@dims @liggitt ptal, only last 3 commits. Rebased onto @deads2k's giant config refactoring https://github.com/kubernetes/kubernetes/pull/36604

sttts commented:
Blocked on https://github.com/kubernetes/kubernetes/pull/36604

Commit:
Turn MaybeGenerateServingCerts into part of ApplySecureServingOptions

This removes all dependencies on Config during cert generation, only operating
on ServerRunOptions. This way we get rid of the repeated call of Config.Complete
and cleanly stratify the GenericApiServer bootstrapping.

Commit:
Use cfssl to create apiserver dev certs and root-ca

Commit:
Update bazel

Commit:
Update bazel

dims commented:
Are we still going to do this @sttts ?

deads2k commented:
> Are we still going to do this @sttts ?

I doubt we're going to solve it like this.  I think its more likely that we demonstrate how to use cfssl to create certs ahead of time.

deads2k commented:
See https://github.com/kubernetes/kubernetes/pull/38409 as an example.

sttts commented:
@deads2k having separate CA and cert is cleaner. It solves the `curl -k` issue. Go is fine with trusting the cert itself. I wouldn't mind to get this in, but it's not critical.

dims commented:
/unassign dims

dims commented:
/unassign @dims
